### PR TITLE
admission: introduce elastic io token exhausted duration metric

### PR DESCRIPTION
**Pull request by aadityasondhi**
This patch adds a new metric `elastic_io_tokens_exhausted_duration.kv`. This is similar to the existing
`admission.granter.io_tokens_exhausted_duration.kv`, but for elastic traffic.

The patch also does some code refactoring to make it easier to use both regular and elastic equivalent metrics.

Informs #121574.

Release note: None

### CODE REVIEW

**Comment by aadityasondhi** [comment]
In DB console:

image

**Review by sumeerbhola — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124078)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/util/admission/granter.go` line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124078#-Nxwl-u_EDBMLjxIYeUt:-Nxwl-u_EDBMLjxIYeUu:b-y0889v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb1730e4073d77008f54888babda17c787e3a3c/pkg/util/admission/granter.go#L471)):*
> ```Go
> 		updateExhaustedTokenDurationLocked(tokenCount, avail[w], sg.coordMu.availableIOTokens[w], settingAvailableTokens,
> 			&sg.exhaustedStart[w], sg.ioTokensExhaustedDurationMetric[w])
> 	}
> ```

This looping and separation of the updateExhausterTokenDurationLocked with all these params is hard to follow. I think the following alternative is simpler:

```
func (sg *kvStoreTokenGranter) subtractTokensLocked(
	count int64, elasticCount int64, settingAvailableTokens bool,
) {
	sg.subtractTokensLockedForWorkClass(admissionpb.RegularWorkClass, count, settingAvailableTokens)
	sg.subtractTokensLockedForWorkClass(admissionpb.ElasticWorkClass, elasticCount, settingAvailableTokens
	if !settingAvailableTokens {
		if count > 0 {
			sg.tokensTakenMetric.Inc(count)
		} else {
			sg.tokensReturnedMetric.Inc(-count)
		}
	}
}

func (sg *kvStoreTokenGranter) subtractTokensLockedForWorkClass(wc admissionpb.WorkClass, count int64, settingAvailableTokens bool) {
	sg.coordMu.availableIOTokens[wc] -= count
	sg.availableTokensMetric[wc].Update(sg.coordMu.availableIOTokens[wc])
	// Inline the code from updateExhaustedTokenDurationLocked.
  ...
}
```

**Review by aadityasondhi — COMMENTED** [review]
TFTR, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124078#-:-NxxD1RnCn7ffwlpkrIc:b-qs6e22)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/util/admission/granter.go` line 471 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/124078#-Nxwl-u_EDBMLjxIYeUt:-NxxCnsY1XHW4mDEFaAZ:bful9ky) ([raw file](https://github.com/cockroachdb/cockroach/blob/4bb1730e4073d77008f54888babda17c787e3a3c/pkg/util/admission/granter.go#L471)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This looping and separation of the updateExhausterTokenDurationLocked with all these params is hard to follow. I think the following alternative is simpler:

```
func (sg *kvStoreTokenGranter) subtractTokensLocked(
	count int64, elasticCount int64, settingAvailableTokens bool,
) {
	sg.subtractTokensLockedForWorkClass(admissionpb.RegularWorkClass, count, settingAvailableTokens)
	sg.subtractTokensLockedForWorkClass(admissionpb.ElasticWorkClass, elasticCount, settingAvailableTokens
	if !settingAvailableTokens {
		if count > 0 {
			sg.tokensTakenMetric.Inc(count)
		} else {
			sg.tokensReturnedMetric.Inc(-count)
		}
	}
}

func (sg *kvStoreTokenGranter) subtractTokensLockedForWorkClass(wc admissionpb.WorkClass, count int64, settingAvailableTokens bool) {
	sg.coordMu.availableIOTokens[wc] -= count
	sg.availableTokensMetric[wc].Update(sg.coordMu.availableIOTokens[wc])
	// Inline the code from updateExhaustedTokenDurationLocked.
  ...
}
```
</blockquote></details>

Done. Thanks for the suggestion, code looks much cleaner (and more aligned with what i was looking to do here).

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/124078#-:-NxxdFdE4QaTLTWFFc--:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi)

**Comment by aadityasondhi** [comment]
bors r=sumeerbhola
